R/return_random_cases.R

```diff
@@ -0,0 +1,82 @@
+#'             This date is checked against mnpvisstartdate in the casenodes table.
+#' @param seed Allows to \code{set.seed} for id sampling.
+#' @export
+#' @details return_random_cases will produce a data.frame that contains random cases from each
+#'          specified centre. This is performed based on a specified seed to retain reproducibilty.
+#'
+#' @examples
+#' # export location
+#' expot_loc <- system.file("extdata", "sT_exports", "longnames",
+#'                          "s_export_CSV-xls_CTU05_long_ref_miss_en_utf8.zip",
+#'                          package = "secuTrialR")
+#' # read export
+#' sT_export <- read_secuTrial(expot_loc)
+#'
+#' # return random cases
+#' return_random_cases(sT_export, percent = 0.25, seed = 1337, date = "2019-03-18",
+#'                     centres = c("Inselspital Bern (RPACK)", "Charité Berlin (RPACK)"))
+#'
+return_random_cases <- function(x, centres = "all", percent = 0.1, date = "1900-01-01", seed = 1) {
+  if (class(x) == "secuTrialdata") {
+     cn_table <- x[[x$export_options$meta_names$casenodes]]
+     ctr_table <- x[[x$export_options$meta_names$centres]]
+     # need add ids because these will be returned
+     if(! x$export_options$add_id) {
+       stop("Please reexport with Add-ID.")
+     }
+     # need centre info
+     if(! x$export_options$centre_info) {
+       stop("Please reexport with Centre information.")
+     }
+     # 0 and 1 can also be excluded because 0 means nothing
+     # and 1 means everything. Both cases are implicitly useless.
+     if (percent <= 0 | percent >= 1) {
+       stop("Please specify a number greater than 0 and smaller than 1.")
+     }
+     # check date format
+     date_parsed <- ymd(date)
+     if (is.na(date_parsed)) {
+       stop("The specified date is not parsable by lubridate::ymd. Please use format 'YYYY-MM-DD'.")
+     }
+
+     if ("all" %in% centres) {
+       centres <- ctr_table$mnpctrname
+     }
+
+     # add centre strings
+     cn_table <- add_centre_col(cn_table, casenodes_table = cn_table, centre_table = ctr_table)
+     # remove cases before specified date
+     removed_date_idx <- which(ymd(cn_table$mnpvisstartdate) <= date_parsed)
+     if(length(removed_date_idx)) {
+       cn_table <- cn_table[-removed_date_idx, ]
+     }
+     # init output data.frame
+     random_cases <- setNames(data.frame(matrix(ncol = 3, nrow = 0)),
+                              c("mnpaid", "centre", "mnpvisstartdate"))
+     for (ctr in centres) {
+       curr_ctr_cn <- cn_table %>% filter(cn_table$centre == ctr)
+       curr_ctr_cn_trunc <- curr_ctr_cn[, c("mnpaid", "centre", "mnpvisstartdate")]
+       case_ids_count <- length(curr_ctr_cn$mnpaid)
+       # sample count / round up
+       sample_count <- ceiling(case_ids_count * percent)
+       # set seed
+       set.seed(seed)
+       # sample
+       random_cases <- rbind(random_cases,
+                             curr_ctr_cn_trunc[sample(nrow(curr_ctr_cn_trunc), sample_count), ])
```

Comment:
Not sure how good a solution this is... this will take the same indices from each center e.g. the 1st, 3rd, 13th... maybe use a counter and increment the seed for each center (use the provided seed for the first center, seed+1 for the second and so on...

Reply:
Isn't the most important thing that it is reproducible? To me the randomness seems pronounced enough but if you feel strongly about this in your capacity as a statistician I will gladly change it.

Reply:
it's still reproducible, as you still have a seed (and a different one for each center). The only place it might not be reproducible would be if a new center (or participant) was added and it ended up in the middle of the list rather than at the end (and that would only be an issue between exports). I would rather have a different set of indices per site. In this case, I'm not actually sure how important the reproducible aspect is... (if it weren't part of the function, the user could still set the seed before running the command and end up with a reproducible result, _with_ different indices per center)

Reply:
Ok, I will do the increment as suggested. Thank you for the comments.

Reply:
For transparency, can you add a note that a different seed is used for each center?

Reply:
"Every centre will use a small variation of this seed." This is not enough?

Reply:
Ah, yes, sorry, i missed that

Reply:
No worries. Thank you for the review.